torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2897,7 +2901,8 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     # reduce_dtype matches the param dtype.
                     param.grad.data = param.grad.data.to(self.mixed_precision.reduce_dtype)
 
-                if self.gradient_predivide_factor > 1:
+                if self.gradient_predivide_factor > 1 and \
+                        self.communication_hook == self._get_default_comm_hook():
```

Comment:
This logic is pretty tricky. The specification this implies is that we only pre-divide the gradient when using the default communication hook. However, how can a user pre-divide the gradient otherwise? My understanding is that it should be done inside the hook.

Would it be better to pre-divide the gradient in the new all-reduce hook for `NO_SHARD` to be consistent with all registered hooks?

Then, the logic looks like:
- If you are using `NO_SHARD`, then the communication hook is *always* responsible for the pre-division, even if it is the default all-reduce hook.
- If you are using another strategy, then FSDP will pre-divide for you since there is no way for you to pre-divide yourself.

(In short, I do not think we should treat the `NO_SHARD` default hook as a special case and only pre-divide for it but not other hooks.)

Reply:
@awgu I see what you mean. @rohan-varma and @zhaojuanmao mentioned earlier (in previous version of this PR), that we should base pre-division logic on the sharding strategy. For this iteration, I based my logic on whether a default hook is picked or not to retain the default behavior. That being said, I use `communication_hook` as a proxy to determine a sharding strategy:  if `communication_hook` is `None` ( which is true for sharded strategies). Then, I believe when communication hooks are implemented for all sharding strategies, pre-division can be done solely in hooks (i.e. get rid of pre-division here).

Reply:
I might be missing something. Why do we not do something like:
```
if self.sharding_strategy != ShardingStrategy.NO_SHARD:
    param.grad.div_(self.gradient_predivide_factor)
```
and then add the `param.grad.div_(self.gradient_predivide_factor)` as well in the default `NO_SHARD` communication hook before the all-reduce?

Does this not preserve the existing behavior as well?

Reply:
Referencing an earlier discussion: https://github.com/pytorch/pytorch/pull/79833#discussion_r900703089

Reply:
I've refactored `all_reduce` to have pre-division. As for general logic in pre-division, I believe even though doing a pre-division based on a sharding strategy  may preserve the behavior,  the first question would be: Why specifically `NO_SHARD` is excluded - a context is needed (i.e. for `NO_SHARD` pre-division is implemented inside of its default hook). If we do pre-division based on a value of `self.communication_hook`, it at least gives a hint, that pre-division is implemented inside of a `self.communication_hook` when it is not `None`. Since at that point for `SHARD_GRAD_OP` and `FULL_SHARD` default hook is `None`, it does give an idea why pre-division is still there in place.

Reply:
> Why specifically `NO_SHARD` is excluded - a context is needed (i.e. for `NO_SHARD` pre-division is implemented inside of its default hook). 

I am not sure I am following this part.

Specifically `NO_SHARD` is excluded because now the communication is refactored into a hook. Logically, everything related to the the gradient communication is _encapsulated_ into the hook. If we implement communication hooks for the other strategies, I expect those hooks will run the pre- and post-division as well.

It does not make sense to me why we would only attribute post-division to be part of FSDP's semantics and not the pre-division. It should be either both pre- and post-division or neither (especially when the two factors are expected to be reciprocals). Right now, I prefer having both in the hook because there may exist communication algorithms for which pre-division does not apply (e.g. the communication algorithm's reduction does not scale with the world size).

To elaborate, do we want to enforce that all hooks have a state and that that state has a `gradient_predivide_factor` attribute? That seems restrictive. Hence, if we do the pre-division only for the default hook, then we are hard coding the fact that the default hook _does_ have a `gradient_predivide_factor` attribute, which is brittle.

Your previous implementation looked like (before the recent commits):
```
if self.gradient_predivide_factor > 1 and \
        self.communication_hook == self._get_default_comm_hook():
    # Average grad by world_size for consistency with PyTorch DDP.
    param.grad.div_(self.gradient_predivide_factor)
```
This actually uses the FSDP instance's `self.gradient_predivide_factor` and not the one from the `NO_SHARD` default hook. If this was corrected, you would see how you have to assume that `hasattr(self.communication_hook, "gradient_predivide_factor")`.

> If we do pre-division based on a value of `self.communication_hook`, it at least gives a hint, that pre-division is implemented inside of a `self.communication_hook` when it is not `None`. Since at that point for `SHARD_GRAD_OP` and `FULL_SHARD` default hook is `None`, it does give an idea why pre-division is still there in place.

To clarify, is this part explaining why now the check for FSDP's pre-division is `if self.communication_hook is None`?
If so, I think that checking `self.communication_hook` against `None` is fine. Would I be correct to say that it is currently equivalent to checking if the sharding strategy is not `NO_SHARD`?

Reply:
@awgu I agree with you on all of your concerns and previous logic was revised. I do agree communication hooks should take care of both pre- and post-division (if any is required). 

> Specifically NO_SHARD is excluded because now the communication is refactored into a hook. 

I was trying to say, that I know this context, but not everyone does.

>Logically, everything related to the the gradient communication is encapsulated into the hook. If we implement communication hooks for the other strategies, I expect those hooks will run the pre- and post-division as well.

I apologize, that my previous version of PR caused a confusion, but I did have the same understanding in mind. I do understand where confusion comes from and I will avoid it in future. 

> Hence, if we do the pre-division only for the default hook, then we are hard coding the fact that the default hook does have a gradient_predivide_factor attribute, which is brittle.

Previous implementation used FSDP instance's `self.gradient_predivide_factor` to pre-divide gradients and then the default hook post-divided them with a recomputed `gradient_postdivide_factor` (stored in its state). This split was unnecessary, I agree. 

Current pre-division for `SHARD_GRAD_OP` and `FULL_SHARD` strategies  is left in place to preserve a correct behavior, because these strategies don't have a default hook implemented for them. My understanding was, that we can fully transfer this pre-division into default hooks once all of them are implemented and we can get rid of this pre-division here (which I failed to explain, I apologize). Getting rid of pre-division at that point, would also introduce changes to a default `all_reduce` hook (i.e. adding pre-division).

> To clarify, is this part explaining why now the check for FSDP's pre-division is if self.communication_hook is None?

I tried my best to explain that.

> Would I be correct to say that it is currently equivalent to checking if the sharding strategy is not NO_SHARD?

Currently, yes.

Reply:
Thanks for the clarifications! One last question:

> Getting rid of pre-division at that point, would also introduce changes to a default all_reduce hook (i.e. adding pre-division).

I thought the default all-reduce hook now (in your PR) has the pre-division built in, so we do not need to add pre-division to it? It seems like we would not need to modify the default all-reduce hook when we add hooks for the other strategies.

Reply:
I was trying to provide explanations for the previous version. Of course, since in current version pre-division is already in `all_reduce` we won't need to add an extra pre-division later, i.e. less things to keep in mind.